apps/studio/src/server/modules/resource/resource.service.ts

```diff
@@ -409,3 +410,183 @@ export const publishResource = async (
+          eb
+            .selectFrom("Resource as s")
+            .innerJoin("resourcePath as rp", "s.parentId", "rp.id")
+            .select([
+              "s.id",
+              "s.title",
+              "s.permalink",
+              "s.parentId",
+              sql<string>`CONCAT(rp."fullPermalink", '/', s.permalink)`.as(
+                "fullPermalink",
+              ),
+            ]),
+        ),
+    )
+    .selectFrom("resourcePath as rp")
+    .select(["rp.id", "rp.title", "rp.fullPermalink"])
+    .where("rp.id", "=", resourceId)
+    .executeTakeFirst()
+
+  return result
+}
+
+const getResourcesWithLastUpdatedAt = ({ siteId }: { siteId: number }) => {
+  return db
+    .selectFrom("Resource")
+    .select([
+      "Resource.id",
+      "Resource.title",
+      "Resource.type",
+      "Resource.parentId",
+      // To handle cases where either the resource or the blob is updated
+      sql`GREATEST("Resource"."updatedAt", "Blob"."updatedAt")`.as(
+        "lastUpdatedAt",
+      ),
+    ])
+    .leftJoin("Blob", "Resource.draftBlobId", "Blob.id")
+    .where("Resource.siteId", "=", siteId)
+}
+
+const getResourcesWithFullPermalink = async ({
+  resources,
+}: {
+  resources: Omit<SearchResultResource, "fullPermalink">[]
+}): Promise<SearchResultResource[]> => {
+  return await Promise.all(
+    resources.map(async (resource) => ({
+      ...resource,
+      fullPermalink: await getWithFullPermalink({
+        resourceId: resource.id,
+      }).then((r) => r?.fullPermalink ?? ""),
+    })),
+  )
+}
+
+export const getSearchResults = async ({
+  siteId,
+  query,
+  offset,
+  limit,
+}: {
+  siteId: number
+  query: string
+  offset: number
+  limit: number
+}): Promise<{
+  totalCount: number | null
+  resources: SearchResultResource[]
+}> => {
+  const searchTerms: string[] = Array.from(
+    new Set(query.trim().toLowerCase().split(/\s+/)),
+  )
+
+  const queriedResources = getResourcesWithLastUpdatedAt({
+    siteId: Number(siteId),
+  })
+    .where("Resource.type", "in", [
+      // only show user-viewable resources (excluding root page, folder meta etc.)
+      ResourceType.Page,
+      ResourceType.Folder,
+      ResourceType.Collection,
+      ResourceType.CollectionLink,
+      ResourceType.CollectionPage,
+    ])
+    .where((eb) =>
+      eb.or([
+        ...searchTerms.map((searchTerm) =>
+          // Match if the search term is at the start of the title
+          eb("Resource.title", "ilike", `${searchTerm}%`).or(
+            // Match if the search term is in the middle of the title (after a space)
+            eb("Resource.title", "ilike", `% ${searchTerm}%`),
+          ),
+        ),
+      ]),
+    )
+
+  // Currently ordered by number of words matched
+  // followed by `lastUpdatedAt` if there's a tie-break
+  let orderedResources = queriedResources
+  if (searchTerms.length > 1) {
+    orderedResources = orderedResources.orderBy(
+      sql`(
+        ${sql.join(
+          searchTerms.map(
```

Comment:
question(non-blocking): I'm not too sure _why_ this works because it seems like we don't sum the number of matches then use that sum for ordering, so the max that each resource will have is 1. If it has been tested and it works then ok!

Reply:
oh, this actually does because for every match for the searchterm we give it a score of `1` then we sum them based on the number of terms
```ts
    orderedResources = orderedResources.orderBy(
      sql`(
        ${sql.join(
          searchTerms.map(
            (searchTerm) =>
              // 1. Match if the search term is at the start of the title
              // 2. Match if the search term is in the middle of the title (after a space)
              sql`
                CASE WHEN "Resource"."title" ILIKE ${searchTerm + "%"} THEN 1 ELSE 0 END +
                CASE WHEN "Resource"."title" ILIKE ${"% " + searchTerm + "%"} THEN 1 ELSE 0 END
              `,
          ),
          sql` + `,
        )}
      ) DESC`,
```

an equivalent SQL query will be
```sql
order by 
  CASE WHEN "Resource"."title" ILIKE 'word1%' THEN 1 ELSE 0 END +
  CASE WHEN "Resource"."title" ILIKE '% word1%' THEN 1 ELSE 0 END +
  CASE WHEN "Resource"."title" ILIKE 'word2%' THEN 1 ELSE 0 END +
  CASE WHEN "Resource"."title" ILIKE '% word2%' THEN 1 ELSE 0 END +
  ...
```